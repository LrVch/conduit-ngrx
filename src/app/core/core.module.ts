import { NgModule, Optional, SkipSelf } from '@angular/core';
import { CommonModule } from '@angular/common';
import { HTTP_INTERCEPTORS } from '@angular/common/http';
import { HttpTokenInterceptor } from './interceptors/http.token.interceptor';

import {
  ApiService,
  ArticlesService,
  AuthGuard,
  CommentsService,
  JwtService,
  ProfilesService,
  TagsService,
  UserService,
  CustomEventManager,
  ScrollService,
  DomUtilService
} from './services';
import { EventManager } from '@angular/platform-browser';

@NgModule({
  imports: [
    CommonModule
  ],
  providers: [
    { provide: HTTP_INTERCEPTORS, useClass: HttpTokenInterceptor, multi: true },
    ApiService,
    ArticlesService,
    AuthGuard,
    CommentsService,
    JwtService,
    ProfilesService,
    TagsService,
    UserService,
    ScrollService,
    DomUtilService,
    { provide: EventManager, useClass: CustomEventManager }
  ],
  declarations: []
})
export class CoreModule {
  constructor (@Optional() @SkipSelf() parentModule: CoreModule) {
    if (parentModule) {
      throw new Error(
        'CoreModule is already loaded. Import it in the AppModule only');
    }
  }
 }
