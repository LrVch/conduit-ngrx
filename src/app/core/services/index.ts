export * from './api.service';
export * from './articles.service';
export * from './auth-guard.service';
export * from './comments.service';
export * from './profiles.service';
export * from './tags.service';
export * from './user.service';
export * from './custom-event-manager';
export * from './main-loader.service';
export * from './scroll.service';
export * from './dom-util.service';
export * from './local-storage.service';
export * from './animations.service';
