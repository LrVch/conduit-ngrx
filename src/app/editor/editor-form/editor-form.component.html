<form [formGroup]="articleForm">
    <fieldset [disabled]="disabled">

      <mat-form-field class="input-full-width">
        <input formControlName="title"
               type="text"
               matInput
               placeholder="Article Title">
      </mat-form-field>

      <mat-form-field class="input-full-width">
        <input formControlName="description"
               type="text"
               matInput
               placeholder="What's this article about?">
      </mat-form-field>

      <mat-form-field class="input-full-width">
        <textarea formControlName="body"
                  matInput
                  rows="8"
                  placeholder="Write your article (in markdown)"></textarea>
      </mat-form-field>

      <mat-form-field class="input-full-width">
        <mat-chip-list #chipList>
          <mat-chip *ngFor="let tag of tags"
                    [selectable]="selectable"
                    [removable]="removable"
                    (removed)="removeTag(tag)">
            {{tag.name}}
            <mat-icon matChipRemove
                      *ngIf="removable">cancel</mat-icon>
          </mat-chip>
          <input placeholder="Enter tags"
                 [matChipInputFor]="chipList"
                 [matChipInputSeparatorKeyCodes]="separatorKeysCodes"
                 [matChipInputAddOnBlur]="addOnBlur"
                 (matChipInputTokenEnd)="addTag($event)">
        </mat-chip-list>
      </mat-form-field>

      <button (click)="submitForm()"
              mat-flat-button
              [disabled]="disabled"
              type="submit"
              color="primary">Publish Article</button>

    </fieldset>
  </form>