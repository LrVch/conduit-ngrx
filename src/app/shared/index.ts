export * from './article/article.component';
export * from './articles-list/articles-list.component';
export * from './content-tabs/content-tabs.component';
export * from './tags/tags.component';
export * from './favorite/favorite.component';
export * from './layout';
export * from './shared.module';
export * from './utils';
export * from './list-error/list-error.component';
export * from './show-authed.directive';
export * from './user-info/user-info.component';
export * from './follow-button/follow-button.component';
export * from './tag-list/tag-list.component';
export * from './comments/comments.component';
export * from './base-from/base-from.component';
