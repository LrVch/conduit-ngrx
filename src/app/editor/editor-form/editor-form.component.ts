import { Component, OnInit, Input, Output, EventEmitter, OnDestroy } from '@angular/core';
import { FormBuilder, Validators } from '@angular/forms';
import { Article } from '@app/core';
import { Tag } from '../editor/editor.component';
import { COMMA, ENTER } from '@angular/cdk/keycodes';
import { BaseFromComponent } from '@app/shared';
import { withLatestFrom, debounceTime, map, startWith, takeUntil, tap, distinctUntilChanged } from 'rxjs/operators';
import { of, Observable, Subject, combineLatest, BehaviorSubject, merge, asyncScheduler } from 'rxjs';

@Component({
  selector: 'app-editor-form',
  templateUrl: './editor-form.component.html'
})
export class EditorFormComponent extends BaseFromComponent implements OnInit, OnDestroy {
  @Input('article') article: Article = {} as Article;
  @Output() onsubmit = new EventEmitter<Article>();
  @Output() wasChanged = new EventEmitter<boolean>();

  initState$ = new BehaviorSubject<{ form: {}, tags: Tag[] }>({ form: {}, tags: [] });
  unsubscribe$ = new Subject<any>();
  tags$ = new BehaviorSubject<Tag[]>([]);
  tagsObs$ = this.tags$.asObservable();

  tags: Tag[] = [];
  tagConfig = {
    selectable: true,
    removable: true,
    addOnBlur: true
  };

  readonly separatorKeysCodes: number[] = [ENTER, COMMA];

  constructor(
    private fb: FormBuilder
  ) {
    super();
  }

  ngOnInit() {
    this.form = this.fb.group({
      title: ['', [Validators.required]],
      description: ['', [Validators.required]],
      body: ['', [Validators.required]]
    });

    if (this.article) {
      this.form.patchValue(this.article);
      this.tags = this.article.tagList.map(tag => ({ name: tag }));
    }

    this.initState$.next({ form: this.form.value, tags: this.tags });
    this.tags$.next(this.tags);

    combineLatest(this.form.valueChanges.pipe(startWith(this.form.value)), this.tagsObs$).pipe(
      map(([form, tags]) => ({ form, tags })),
      tap(console.log),
      withLatestFrom(this.initState$),
      debounceTime(200),
      map(([form, initState]) => JSON.stringify(form) === JSON.stringify(initState)),
      startWith(true),
      takeUntil(this.unsubscribe$)
    ).subscribe(value => {
      this.wasChanged.emit(!value);
    });
  }

  ngOnDestroy() {
    this.unsubscribe$.next();
    this.unsubscribe$.complete();
    this.tags$.complete();
  }

  addTag($event) {
    const tagName = $event.value;
    const input = $event.input;

    if ((tagName || '').trim() && this.tags.indexOf(tagName) < 0) {
      this.tags = [...this.tags, { name: tagName.trim() }];
      this.tags$.next(this.tags);
    }

    input.value = '';
  }

  removeTag(tagToRemove: Tag) {
    this.tags = this.tags.filter(tag => tag.name !== tagToRemove.name);
    this.tags$.next(this.tags);
  }

  submit() {
    if (this.form.valid) {
      const article = Object.assign({}, this.article, this.form.value);
      article.tagList = this.tags.map(tag => tag.name);

      this.initState$.next({ form: this.form.value, tags: this.tags });
      this.onsubmit.emit(article);
    } else {
      this.validateAllFormFields(this.form);
    }
  }

  get titleControl() {
    return this.form.get('title');
  }

  get descriptionControl() {
    return this.form.get('description');
  }

  get bodyControl() {
    return this.form.get('body');
  }

  get requiredTitle() {
    return this.titleControl.hasError('required');
  }

  get requiredDescription() {
    return this.descriptionControl.hasError('required');
  }

  get requiredBody() {
    return this.bodyControl.hasError('required');
  }
}
