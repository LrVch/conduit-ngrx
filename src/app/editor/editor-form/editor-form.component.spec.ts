import { async, ComponentFixture, fakeAsync, tick } from '@angular/core/testing';

import { ConfigureFn, configureTests, getUser, getArticle } from '@app/lib/testing';
import { EditorFormComponent } from './editor-form.component';
import { DebugElement } from '@angular/core';
import { ReactiveFormsModule, FormsModule } from '@angular/forms';
import { ScrollService, DomUtilService, Article } from '@app/core';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';
import { By } from '@angular/platform-browser';
import { PasswordStrengthComponent, MaterialModule, AccentOnInvalidFromFieldDirective } from '@app/shared';

describe('EditorFormComponent', () => {
    let component: EditorFormComponent;
    let fixture: ComponentFixture<EditorFormComponent>;
    let de: DebugElement;
    let article: Article;

    beforeEach(
        async(() => {
            article = getArticle();
            const configure: ConfigureFn = testBed => {
                testBed.configureTestingModule({
                    declarations: [EditorFormComponent, AccentOnInvalidFromFieldDirective, PasswordStrengthComponent],
                    imports: [FormsModule, ReactiveFormsModule, MaterialModule, NoopAnimationsModule],
                    providers: [ScrollService, DomUtilService]
                });
            };

            configureTests(configure).then(testBed => {
                fixture = testBed.createComponent(EditorFormComponent);
                component = fixture.componentInstance;
                component.article = article;
                de = fixture.debugElement;
                fixture.detectChanges();

                spyOn(component.unsubscribe$, 'next').and.callThrough();
                spyOn(component.unsubscribe$, 'complete').and.callThrough();
                spyOn(component.tags$, 'complete').and.callThrough();
            });
        })
    );

    it('should create', () => {
        expect(component).toBeTruthy();
    });

    it('should has "submitting" props', () => {
        expect(component.submitting).toBe(false);
    });

    it('should create from with 3 controls', () => {
        const form = component.form;

        expect(form.contains('title')).toBeTruthy();
        expect(form.contains('description')).toBeTruthy();
        expect(form.contains('body')).toBeTruthy();
    });

    it('title control should be required', () => {
        const form = component.form;
        const title = form.get('title');

        title.setValue('');

        expect(title.valid).toBeFalsy();
    });

    it('description control should be required', () => {
        const form = component.form;
        const description = form.get('description');

        description.setValue('');

        expect(description.valid).toBeFalsy();
    });

    it('body control should be required', () => {
        const form = component.form;
        const body = form.get('body');

        body.setValue('');

        expect(body.valid).toBeFalsy();
    });

    it('should show error message', () => {
        component.form.get('title').setValue('');
        component.form.get('description').setValue('');
        component.form.get('body').setValue('');

        component.submit();

        fixture.detectChanges();

        (<any>expect(fixture)).toMatchSnapshot();
    });

    it('should disable button', () => {
        component.isSubmitting = true;
        fixture.detectChanges();
        (<any>expect(fixture)).toMatchSnapshot();
    });

    it('should disable form', () => {
        component.isSubmitting = true;

        expect(component.form.status).toBe('DISABLED');
    });

    it('should raise "onsubmit" when form valid on submit event and pass user data', () => {
        let title = '';
        let description = '';
        let body = '';
        let tagList = [''];
        const expectedTitle = 'title';
        const expectedDescription = 'description';
        const expectedBody = 'body';
        const expectedTags = [{ name: 'one' }];

        component.form.get('title').setValue(expectedTitle);
        component.form.get('description').setValue(expectedDescription);
        component.form.get('body').setValue(expectedBody);
        component.tags = expectedTags;

        component.onsubmit.subscribe((res: Article) => {
            ({ title, description, body, tagList } = res);
        });

        fixture.detectChanges();
        const formDe = de.query(By.css('form'));

        formDe.triggerEventHandler('ngSubmit', null);

        expect(title).toBe(expectedTitle);
        expect(description).toBe(expectedDescription);
        expect(body).toBe(expectedBody);
        expect(tagList).toEqual(expectedTags.map(tag => tag.name));
    });

    it('shouldn\'t raise "onsubmit" when form invalid', () => {
        let wasEmited = false;

        component.form.get('title').setValue('');

        component.onsubmit.subscribe(c => wasEmited = true);

        fixture.detectChanges();
        const formDe = de.query(By.css('form'));

        formDe.triggerEventHandler('ngSubmit', null);

        expect(wasEmited).toBe(false);
    });

    it('should invoke validateAllFormFields method when form invalid', () => {
        const spy = jest.spyOn(component, 'validateAllFormFields');

        component.form.get('title').setValue('');

        fixture.detectChanges();
        const formDe = de.query(By.css('form'));
        formDe.triggerEventHandler('ngSubmit', null);

        expect(spy).toHaveBeenCalled();
    });

    it('should raise "wasChanged" when form data changes', fakeAsync(() => {
        spyOn(component.wasChanged, 'emit');
        component.ngOnInit();

        component.form.get('title').setValue('titlefds');

        tick(200);

        expect(component.wasChanged.emit).toHaveBeenCalledWith(true);

        component.form.get('title').setValue('title');

        tick(200);

        expect(component.wasChanged.emit).toHaveBeenCalledWith(false);
    }));

    it('should handle ngOnDestroy', () => {
        component.ngOnDestroy();

        expect(component.unsubscribe$.next).toHaveBeenCalled();
        expect(component.unsubscribe$.complete).toHaveBeenCalled();
        expect(component.tags$.complete).toHaveBeenCalled();
    });

    it('should add tag', () => {
        const input = {
            value: ''
        };
        const event = {
            value: 'tag',
            input
        };
        const expected = [{ name: 'list' }, { name: 'tag' }];

        component.addTag(event);

        expect(component.tags).toEqual(expected);

        component.tags$.subscribe(res => {
            expect(res).toEqual(expected);
        });

    });

    it('should remove tag', () => {
        const expected = [];

        component.removeTag({ name: 'list' });

        expect(component.tags).toEqual(expected);

        component.tags$.subscribe(res => {
            expect(res).toEqual(expected);
        });

    });
});
