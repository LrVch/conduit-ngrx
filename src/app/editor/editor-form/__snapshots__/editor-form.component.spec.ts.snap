// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditorFormComponent should disable button 1`] = `
<app-editor-form
  article={[Function Object]}
  fb={[Function FormBuilder]}
  form={[Function FormGroup]}
  initState$={[Function BehaviorSubject]}
  onsubmit={[Function EventEmitter]}
  separatorKeysCodes={[Function Array]}
  submitting={[Function Boolean]}
  tagConfig={[Function Object]}
  tags={[Function Array]}
  tags$={[Function BehaviorSubject]}
  tagsObs$={[Function Observable]}
  unsubscribe$={[Function Subject]}
  wasChanged={[Function EventEmitter]}
>
  <form
    class="ng-untouched ng-pristine"
    ng-reflect-app-accent-on-invalid-from-field="[object Object]"
    ng-reflect-form="[object Object]"
    novalidate=""
  >
    <fieldset
      disabled=""
    >
      <mat-form-field
        class="input-full-width mat-form-field ng-tns-c28-0 mat-primary mat-form-field-type-mat-input mat-form-field-appearance-legacy mat-form-field-can-float mat-form-field-should-float mat-form-field-hide-placeholder ng-untouched ng-pristine _mat-animation-noopable mat-form-field-disabled"
      >
        <div
          class="mat-form-field-wrapper"
        >
          <div
            class="mat-form-field-flex"
          >
            
            
            <div
              class="mat-form-field-infix"
            >
              <input
                aria-invalid="false"
                aria-required="false"
                class="mat-input-element mat-form-field-autofill-control cdk-text-field-autofill-monitored ng-untouched ng-pristine"
                disabled=""
                formcontrolname="title"
                id="mat-input-21"
                matinput=""
                ng-reflect-name="title"
                ng-reflect-placeholder="Article Title"
                ng-reflect-type="text"
                placeholder="Article Title"
                type="text"
              />
              <span
                class="mat-form-field-label-wrapper"
              >
                
                <label
                  aria-owns="mat-input-21"
                  class="mat-form-field-label ng-tns-c28-0 ng-star-inserted"
                  for="mat-input-21"
                  id="mat-form-field-label-57"
                  ng-reflect-disabled="true"
                  ng-reflect-ng-switch="false"
                >
                  
                  
                  Article Title
                  
                  
                </label>
              </span>
            </div>
            
          </div>
          
          <div
            class="mat-form-field-underline ng-tns-c28-0 ng-star-inserted"
          >
            <span
              class="mat-form-field-ripple"
            />
          </div>
          <div
            class="mat-form-field-subscript-wrapper"
            ng-reflect-ng-switch="hint"
          >
            
            
            <div
              class="mat-form-field-hint-wrapper ng-tns-c28-0 ng-trigger ng-trigger-transitionMessages ng-star-inserted"
              style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
            >
              
              <div
                class="mat-form-field-hint-spacer"
              />
            </div>
          </div>
        </div>
      </mat-form-field>
      <mat-form-field
        class="input-full-width mat-form-field ng-tns-c28-1 mat-primary mat-form-field-type-mat-input mat-form-field-appearance-legacy mat-form-field-can-float mat-form-field-should-float mat-form-field-hide-placeholder ng-untouched ng-pristine _mat-animation-noopable mat-form-field-disabled"
      >
        <div
          class="mat-form-field-wrapper"
        >
          <div
            class="mat-form-field-flex"
          >
            
            
            <div
              class="mat-form-field-infix"
            >
              <input
                aria-invalid="false"
                aria-required="false"
                class="mat-input-element mat-form-field-autofill-control cdk-text-field-autofill-monitored ng-untouched ng-pristine"
                disabled=""
                formcontrolname="description"
                id="mat-input-22"
                matinput=""
                ng-reflect-name="description"
                ng-reflect-placeholder="What's this article about?"
                ng-reflect-type="text"
                placeholder="What's this article about?"
                type="text"
              />
              <span
                class="mat-form-field-label-wrapper"
              >
                
                <label
                  aria-owns="mat-input-22"
                  class="mat-form-field-label ng-tns-c28-1 ng-star-inserted"
                  for="mat-input-22"
                  id="mat-form-field-label-59"
                  ng-reflect-disabled="true"
                  ng-reflect-ng-switch="false"
                >
                  
                  
                  What's this article about?
                  
                  
                </label>
              </span>
            </div>
            
          </div>
          
          <div
            class="mat-form-field-underline ng-tns-c28-1 ng-star-inserted"
          >
            <span
              class="mat-form-field-ripple"
            />
          </div>
          <div
            class="mat-form-field-subscript-wrapper"
            ng-reflect-ng-switch="hint"
          >
            
            
            <div
              class="mat-form-field-hint-wrapper ng-tns-c28-1 ng-trigger ng-trigger-transitionMessages ng-star-inserted"
              style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
            >
              
              <div
                class="mat-form-field-hint-spacer"
              />
            </div>
          </div>
        </div>
      </mat-form-field>
      <mat-form-field
        class="input-full-width mat-form-field ng-tns-c28-2 mat-primary mat-form-field-type-mat-input mat-form-field-appearance-legacy mat-form-field-can-float mat-form-field-should-float mat-form-field-hide-placeholder ng-untouched ng-pristine _mat-animation-noopable mat-form-field-disabled"
      >
        <div
          class="mat-form-field-wrapper"
        >
          <div
            class="mat-form-field-flex"
          >
            
            
            <div
              class="mat-form-field-infix"
            >
              <textarea
                aria-invalid="false"
                aria-required="false"
                class="mat-input-element mat-form-field-autofill-control cdk-text-field-autofill-monitored ng-untouched ng-pristine"
                disabled=""
                formcontrolname="body"
                id="mat-input-23"
                matinput=""
                ng-reflect-name="body"
                ng-reflect-placeholder="Write your article (in markdow"
                placeholder="Write your article (in markdown)"
                rows="8"
              />
              <span
                class="mat-form-field-label-wrapper"
              >
                
                <label
                  aria-owns="mat-input-23"
                  class="mat-form-field-label ng-tns-c28-2 ng-star-inserted"
                  for="mat-input-23"
                  id="mat-form-field-label-61"
                  ng-reflect-disabled="true"
                  ng-reflect-ng-switch="false"
                >
                  
                  
                  Write your article (in markdown)
                  
                  
                </label>
              </span>
            </div>
            
          </div>
          
          <div
            class="mat-form-field-underline ng-tns-c28-2 ng-star-inserted"
          >
            <span
              class="mat-form-field-ripple"
            />
          </div>
          <div
            class="mat-form-field-subscript-wrapper"
            ng-reflect-ng-switch="hint"
          >
            
            
            <div
              class="mat-form-field-hint-wrapper ng-tns-c28-2 ng-trigger ng-trigger-transitionMessages ng-star-inserted"
              style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
            >
              
              <div
                class="mat-form-field-hint-spacer"
              />
            </div>
          </div>
        </div>
      </mat-form-field>
      <mat-form-field
        class="input-full-width mat-form-field ng-tns-c28-3 mat-primary mat-form-field-type-mat-chip-list mat-form-field-appearance-legacy mat-form-field-can-float mat-form-field-should-float mat-form-field-hide-placeholder _mat-animation-noopable"
      >
        <div
          class="mat-form-field-wrapper"
        >
          <div
            class="mat-form-field-flex"
          >
            
            
            <div
              class="mat-form-field-infix"
            >
              <mat-chip-list
                aria-disabled="false"
                aria-invalid="false"
                aria-multiselectable="false"
                aria-orientation="horizontal"
                aria-required="false"
                class="mat-chip-list"
                id="mat-chip-list-7"
                role="listbox"
                tabindex="0"
              >
                <div
                  class="mat-chip-list-wrapper"
                >
                  
                  <mat-chip
                    aria-disabled="false"
                    aria-selected="false"
                    class="mat-chip mat-primary mat-standard-chip mat-chip-with-trailing-icon ng-star-inserted"
                    ng-reflect-removable="true"
                    ng-reflect-selectable="true"
                    role="option"
                    tabindex="-1"
                  >
                     list 
                    
                    <mat-icon
                      aria-hidden="true"
                      class="mat-chip-remove mat-chip-trailing-icon mat-icon material-icons ng-star-inserted"
                      matchipremove=""
                      role="img"
                    >
                      cancel
                    </mat-icon>
                  </mat-chip>
                  <input
                    class="mat-chip-input mat-input-element"
                    id="mat-chip-list-input-7"
                    ng-reflect-add-on-blur="true"
                    ng-reflect-chip-list="[object Object]"
                    ng-reflect-placeholder="Enter tags"
                    ng-reflect-separator-key-codes="13,188"
                    placeholder="Enter tags"
                  />
                </div>
              </mat-chip-list>
              <span
                class="mat-form-field-label-wrapper"
              >
                
                <label
                  aria-owns="mat-chip-list-input-7"
                  class="mat-form-field-label ng-tns-c28-3 ng-star-inserted"
                  for="mat-chip-list-input-7"
                  id="mat-form-field-label-63"
                  ng-reflect-disabled="true"
                  ng-reflect-ng-switch="false"
                >
                  
                  
                  Enter tags
                  
                  
                </label>
              </span>
            </div>
            
          </div>
          
          <div
            class="mat-form-field-underline ng-tns-c28-3 ng-star-inserted"
          >
            <span
              class="mat-form-field-ripple"
            />
          </div>
          <div
            class="mat-form-field-subscript-wrapper"
            ng-reflect-ng-switch="hint"
          >
            
            
            <div
              class="mat-form-field-hint-wrapper ng-tns-c28-3 ng-trigger ng-trigger-transitionMessages ng-star-inserted"
              style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
            >
              
              <div
                class="mat-form-field-hint-spacer"
              />
            </div>
          </div>
        </div>
      </mat-form-field>
      <button
        class="mat-flat-button mat-primary _mat-animation-noopable"
        color="primary"
        disabled=""
        mat-flat-button=""
        ng-reflect-color="primary"
        ng-reflect-disabled="true"
        type="submit"
      >
        <span
          class="mat-button-wrapper"
        >
          Publish Article
        </span>
        <div
          class="mat-button-ripple mat-ripple"
          matripple=""
          ng-reflect-centered="false"
          ng-reflect-disabled="true"
          ng-reflect-trigger="[object HTMLButtonElement]"
        />
        <div
          class="mat-button-focus-overlay"
        />
      </button>
    </fieldset>
  </form>
</app-editor-form>
`;

exports[`EditorFormComponent should show error message 1`] = `
<app-editor-form
  article={[Function Object]}
  fb={[Function FormBuilder]}
  form={[Function FormGroup]}
  initState$={[Function BehaviorSubject]}
  onsubmit={[Function EventEmitter]}
  separatorKeysCodes={[Function Array]}
  submitting="false"
  tagConfig={[Function Object]}
  tags={[Function Array]}
  tags$={[Function BehaviorSubject]}
  tagsObs$={[Function Observable]}
  unsubscribe$={[Function Subject]}
  wasChanged={[Function EventEmitter]}
>
  <form
    class="ng-untouched ng-pristine ng-invalid"
    ng-reflect-app-accent-on-invalid-from-field="[object Object]"
    ng-reflect-form="[object Object]"
    novalidate=""
  >
    <fieldset>
      <mat-form-field
        class="input-full-width mat-form-field ng-tns-c24-0 mat-primary mat-form-field-type-mat-input mat-form-field-appearance-legacy mat-form-field-can-float mat-form-field-hide-placeholder ng-pristine _mat-animation-noopable mat-form-field-invalid ng-touched ng-invalid"
      >
        <div
          class="mat-form-field-wrapper"
        >
          <div
            class="mat-form-field-flex"
          >
            
            
            <div
              class="mat-form-field-infix"
            >
              <input
                aria-describedby="mat-error-0"
                aria-invalid="true"
                aria-required="false"
                class="mat-input-element mat-form-field-autofill-control cdk-text-field-autofill-monitored ng-pristine ng-touched ng-invalid"
                formcontrolname="title"
                id="mat-input-18"
                matinput=""
                ng-reflect-name="title"
                ng-reflect-placeholder="Article Title"
                ng-reflect-type="text"
                placeholder="Article Title"
                type="text"
              />
              <span
                class="mat-form-field-label-wrapper"
              >
                
                <label
                  aria-owns="mat-input-18"
                  class="mat-form-field-label ng-tns-c24-0 ng-star-inserted mat-empty mat-form-field-empty"
                  for="mat-input-18"
                  id="mat-form-field-label-49"
                  ng-reflect-disabled="true"
                  ng-reflect-ng-switch="false"
                >
                  
                  
                  Article Title
                  
                  
                </label>
              </span>
            </div>
            
          </div>
          
          <div
            class="mat-form-field-underline ng-tns-c24-0 ng-star-inserted"
          >
            <span
              class="mat-form-field-ripple"
            />
          </div>
          <div
            class="mat-form-field-subscript-wrapper"
            ng-reflect-ng-switch="error"
          >
            
            <div
              class="ng-tns-c24-0 ng-trigger ng-trigger-transitionMessages ng-star-inserted ng-animating"
              style=""
            >
              
              <mat-error
                class="mat-error ng-star-inserted"
                id="mat-error-0"
                role="alert"
                style=""
              >
                 Title is 
                <strong>
                  required
                </strong>
              </mat-error>
            </div>
            
          </div>
        </div>
      </mat-form-field>
      <mat-form-field
        class="input-full-width mat-form-field ng-tns-c24-1 mat-primary mat-form-field-type-mat-input mat-form-field-appearance-legacy mat-form-field-can-float mat-form-field-hide-placeholder ng-pristine _mat-animation-noopable mat-form-field-invalid ng-touched ng-invalid"
      >
        <div
          class="mat-form-field-wrapper"
        >
          <div
            class="mat-form-field-flex"
          >
            
            
            <div
              class="mat-form-field-infix"
            >
              <input
                aria-describedby="mat-error-1"
                aria-invalid="true"
                aria-required="false"
                class="mat-input-element mat-form-field-autofill-control cdk-text-field-autofill-monitored ng-pristine ng-touched ng-invalid"
                formcontrolname="description"
                id="mat-input-19"
                matinput=""
                ng-reflect-name="description"
                ng-reflect-placeholder="What's this article about?"
                ng-reflect-type="text"
                placeholder="What's this article about?"
                type="text"
              />
              <span
                class="mat-form-field-label-wrapper"
              >
                
                <label
                  aria-owns="mat-input-19"
                  class="mat-form-field-label ng-tns-c24-1 ng-star-inserted mat-empty mat-form-field-empty"
                  for="mat-input-19"
                  id="mat-form-field-label-51"
                  ng-reflect-disabled="true"
                  ng-reflect-ng-switch="false"
                >
                  
                  
                  What's this article about?
                  
                  
                </label>
              </span>
            </div>
            
          </div>
          
          <div
            class="mat-form-field-underline ng-tns-c24-1 ng-star-inserted"
          >
            <span
              class="mat-form-field-ripple"
            />
          </div>
          <div
            class="mat-form-field-subscript-wrapper"
            ng-reflect-ng-switch="error"
          >
            
            <div
              class="ng-tns-c24-1 ng-trigger ng-trigger-transitionMessages ng-star-inserted ng-animating"
              style=""
            >
              
              <mat-error
                class="mat-error ng-star-inserted"
                id="mat-error-1"
                role="alert"
                style=""
              >
                 Description is 
                <strong>
                  required
                </strong>
              </mat-error>
            </div>
            
          </div>
        </div>
      </mat-form-field>
      <mat-form-field
        class="input-full-width mat-form-field ng-tns-c24-2 mat-primary mat-form-field-type-mat-input mat-form-field-appearance-legacy mat-form-field-can-float mat-form-field-hide-placeholder ng-pristine _mat-animation-noopable mat-form-field-invalid ng-touched ng-invalid"
      >
        <div
          class="mat-form-field-wrapper"
        >
          <div
            class="mat-form-field-flex"
          >
            
            
            <div
              class="mat-form-field-infix"
            >
              <textarea
                aria-describedby="mat-error-2"
                aria-invalid="true"
                aria-required="false"
                class="mat-input-element mat-form-field-autofill-control cdk-text-field-autofill-monitored ng-pristine ng-touched ng-invalid"
                formcontrolname="body"
                id="mat-input-20"
                matinput=""
                ng-reflect-name="body"
                ng-reflect-placeholder="Write your article (in markdow"
                placeholder="Write your article (in markdown)"
                rows="8"
              />
              <span
                class="mat-form-field-label-wrapper"
              >
                
                <label
                  aria-owns="mat-input-20"
                  class="mat-form-field-label ng-tns-c24-2 ng-star-inserted mat-empty mat-form-field-empty"
                  for="mat-input-20"
                  id="mat-form-field-label-53"
                  ng-reflect-disabled="true"
                  ng-reflect-ng-switch="false"
                >
                  
                  
                  Write your article (in markdown)
                  
                  
                </label>
              </span>
            </div>
            
          </div>
          
          <div
            class="mat-form-field-underline ng-tns-c24-2 ng-star-inserted"
          >
            <span
              class="mat-form-field-ripple"
            />
          </div>
          <div
            class="mat-form-field-subscript-wrapper"
            ng-reflect-ng-switch="error"
          >
            
            <div
              class="ng-tns-c24-2 ng-trigger ng-trigger-transitionMessages ng-star-inserted ng-animating"
              style=""
            >
              
              <mat-error
                class="mat-error ng-star-inserted"
                id="mat-error-2"
                role="alert"
                style=""
              >
                 Body is 
                <strong>
                  required
                </strong>
              </mat-error>
            </div>
            
          </div>
        </div>
      </mat-form-field>
      <mat-form-field
        class="input-full-width mat-form-field ng-tns-c24-3 mat-primary mat-form-field-type-mat-chip-list mat-form-field-appearance-legacy mat-form-field-can-float mat-form-field-should-float mat-form-field-hide-placeholder _mat-animation-noopable"
      >
        <div
          class="mat-form-field-wrapper"
        >
          <div
            class="mat-form-field-flex"
          >
            
            
            <div
              class="mat-form-field-infix"
            >
              <mat-chip-list
                aria-disabled="false"
                aria-invalid="false"
                aria-multiselectable="false"
                aria-orientation="horizontal"
                aria-required="false"
                class="mat-chip-list"
                id="mat-chip-list-6"
                role="listbox"
                tabindex="0"
              >
                <div
                  class="mat-chip-list-wrapper"
                >
                  
                  <mat-chip
                    aria-disabled="false"
                    aria-selected="false"
                    class="mat-chip mat-primary mat-standard-chip mat-chip-with-trailing-icon ng-star-inserted"
                    ng-reflect-removable="true"
                    ng-reflect-selectable="true"
                    role="option"
                    tabindex="-1"
                  >
                     list 
                    
                    <mat-icon
                      aria-hidden="true"
                      class="mat-chip-remove mat-chip-trailing-icon mat-icon material-icons ng-star-inserted"
                      matchipremove=""
                      role="img"
                    >
                      cancel
                    </mat-icon>
                  </mat-chip>
                  <input
                    class="mat-chip-input mat-input-element"
                    id="mat-chip-list-input-6"
                    ng-reflect-add-on-blur="true"
                    ng-reflect-chip-list="[object Object]"
                    ng-reflect-placeholder="Enter tags"
                    ng-reflect-separator-key-codes="13,188"
                    placeholder="Enter tags"
                  />
                </div>
              </mat-chip-list>
              <span
                class="mat-form-field-label-wrapper"
              >
                
                <label
                  aria-owns="mat-chip-list-input-6"
                  class="mat-form-field-label ng-tns-c24-3 ng-star-inserted"
                  for="mat-chip-list-input-6"
                  id="mat-form-field-label-55"
                  ng-reflect-disabled="true"
                  ng-reflect-ng-switch="false"
                >
                  
                  
                  Enter tags
                  
                  
                </label>
              </span>
            </div>
            
          </div>
          
          <div
            class="mat-form-field-underline ng-tns-c24-3 ng-star-inserted"
          >
            <span
              class="mat-form-field-ripple"
            />
          </div>
          <div
            class="mat-form-field-subscript-wrapper"
            ng-reflect-ng-switch="hint"
          >
            
            
            <div
              class="mat-form-field-hint-wrapper ng-tns-c24-3 ng-trigger ng-trigger-transitionMessages ng-star-inserted"
              style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
            >
              
              <div
                class="mat-form-field-hint-spacer"
              />
            </div>
          </div>
        </div>
      </mat-form-field>
      <button
        class="mat-flat-button mat-primary _mat-animation-noopable"
        color="primary"
        mat-flat-button=""
        ng-reflect-color="primary"
        ng-reflect-disabled="false"
        type="submit"
      >
        <span
          class="mat-button-wrapper"
        >
          Publish Article
        </span>
        <div
          class="mat-button-ripple mat-ripple"
          matripple=""
          ng-reflect-centered="false"
          ng-reflect-disabled="false"
          ng-reflect-trigger="[object HTMLButtonElement]"
        />
        <div
          class="mat-button-focus-overlay"
        />
      </button>
    </fieldset>
  </form>
</app-editor-form>
`;
